File: playnite_page/webdav.py
"""Upload opzionale del file generato su un server WebDAV.

Variabili d'ambiente attese (in un file .env nella cartella di lavoro
o già presenti nell'ambiente):

    WEBDAV_URL       URL del server WebDAV (obbligatoria)
    WEBDAV_USERNAME  utente (obbligatoria)
    WEBDAV_PASSWORD  password (obbligatoria)
    WEBDAV_PATH      percorso remoto del file (opzionale, default: nome del file)
"""
from __future__ import annotations

import os
from pathlib import Path


class WebDAVConfigError(RuntimeError):
    """Sollevata quando mancano variabili d'ambiente necessarie per l'upload."""


def upload_webdav(file_path: Path) -> None:
    from dotenv import load_dotenv
    from webdav3.client import Client

    load_dotenv()

    hostname = os.getenv("WEBDAV_URL")
    username = os.getenv("WEBDAV_USERNAME")
    password = os.getenv("WEBDAV_PASSWORD")

    missing = [name for name, val in (
        ("WEBDAV_URL", hostname),
        ("WEBDAV_USERNAME", username),
        ("WEBDAV_PASSWORD", password),
    ) if not val]
    if missing:
        raise WebDAVConfigError(
            "Variabili d'ambiente mancanti per l'upload WebDAV: " + ", ".join(missing)
        )

    client = Client({
        "webdav_hostname": hostname,
        "webdav_login": username,
        "webdav_password": password,
    })
    client.verify = True  # Usa False solo con certificati SSL autofirmati

    remote_path = os.getenv("WEBDAV_PATH", "").lstrip("/") or file_path.name

    try:
        client.upload_sync(remote_path=remote_path, local_path=str(file_path))
        print(f"✅ Upload completato su WebDAV: {remote_path}")
    except Exception as exc:  # noqa: BLE001 - vogliamo un messaggio comprensibile all'utente
        print(f"❌ Errore upload WebDAV: {exc}")

File: webdav.go
package main

import (
	"bufio"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

// loadDotEnv legge un file .env molto semplice (KEY=VALUE per riga) e
// imposta le variabili d'ambiente di conseguenza, solo se non già presenti.
// python-dotenv fa lo stesso lavoro in una riga (load_dotenv()); qui lo
// scriviamo a mano perché è un buon esercizio di I/O testuale in Go
// e ci evita una dipendenza esterna per una cosa così piccola.
func loadDotEnv(path string) {
	f, err := os.Open(path)
	if err != nil {
		return // .env assente: non è un errore, semplicemente non c'è nulla da caricare
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		key, value, found := strings.Cut(line, "=")
		if !found {
			continue
		}
		key = strings.TrimSpace(key)
		value = strings.Trim(strings.TrimSpace(value), `"'`)
		if os.Getenv(key) == "" {
			os.Setenv(key, value)
		}
	}
}

// uploadWebDAV carica il file generato sul server WebDAV configurato
// tramite variabili d'ambiente (o file .env nella cartella corrente).
func uploadWebDAV(filePath string) error {
	loadDotEnv(".env")

	hostname := os.Getenv("WEBDAV_URL")
	username := os.Getenv("WEBDAV_USERNAME")
	password := os.Getenv("WEBDAV_PASSWORD")

	var missing []string
	for name, val := range map[string]string{
		"WEBDAV_URL":      hostname,
		"WEBDAV_USERNAME": username,
		"WEBDAV_PASSWORD": password,
	} {
		if val == "" {
			missing = append(missing, name)
		}
	}
	if len(missing) > 0 {
		return fmt.Errorf("variabili d'ambiente mancanti: %s", strings.Join(missing, ", "))
	}

	remotePath := os.Getenv("WEBDAV_PATH")
	if remotePath == "" {
		remotePath = filepath.Base(filePath)
	}

	data, err := os.ReadFile(filePath)
	if err != nil {
		return err
	}

	url := strings.TrimRight(hostname, "/") + "/" + strings.TrimLeft(remotePath, "/")

	req, err := http.NewRequest(http.MethodPut, url, strings.NewReader(string(data)))
	if err != nil {
		return err
	}
	req.SetBasicAuth(username, password)
	req.Header.Set("Content-Type", "text/html")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// WebDAV risponde 201 (Created) per un file nuovo, 204 (No Content) per un overwrite.
	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusNoContent {
		return fmt.Errorf("il server ha risposto con status %d", resp.StatusCode)
	}

	fmt.Println("✅ Upload completato su WebDAV:", remotePath)
	return nil
}
